librad/src/meta/entity/data.rs

```diff
@@ -55,7 +57,50 @@ where
     ordered.serialize(serializer)
 }
 
-/// Raw data for an `Entity
+/// The different kinds of `Entity` types
+#[derive(Clone, Copy, PartialEq, Eq, Deserialize, Serialize, Debug)]
+pub enum EntityKind {
+    User,
+    Project,
+}
+
+/// Information payload for all possible `Entity` types
+///
+/// This is used so we can deserialize `EntityData` without knowing the actual
+/// entity type: deserializing as `EntityData<EntityInfo>` the `info` property
+/// will contain this payload
+///
+/// Care has been taken to ensure that the same data can be deserialized both
+/// "generically" (as `EntityData<EntityInfo>`) and using its specific type
+#[derive(Clone, Deserialize, Serialize, Debug)]
+#[serde(untagged)]
+pub enum EntityInfo {
+    User(UserInfo),
+    Project(ProjectInfo),
+}
+
+pub trait EntityInfoExt: Sized {
```

Comment:
I feel like we don't need a trait for this once we remove `EntityKind`. The `check_invariants` function can just be a method on `EntityInfo`.

Reply:
But we need the trait for `check_invariants`, otherwise we cannot blindly invoke it on `UserInfo`, `ProjectInfo` or `EntityInfo`.
It was one of the simplifications brought by this refactoring (no more `EntityBuilder`).

Reply:
Sure you can if you add `check_invariants` as a method to `EntityInfo`. Then they all implement it. Ez pz

Reply:
What I meant is that given a generic `Entity<T,_>` without a trait you cannot use `check_invariants` because all `T`s do implement it but without the `where T: EntityInfoExt` you cannot invoke it.

Or maybe I am missing something, it's late...

Reply:
:facepalm: my bad. Forgot about the `T` on `Entity`